### PR TITLE
b/221072669: Remove envoy runtime flag preserve_downstream_scheme

### PR DESCRIPTION
**Pull request by qiwzhang**
Signed-off-by: Wayne Zhang <qiwzhang@google.com>

Envoy coming release 1.22 will drop this runtime flag  `preserve_downstream_scheme`.  Currently, we are using this flag to work around a Cloud-Run bug: it sets scheme to "http" if the flag `--use_http2`  is used in Cloud-Run deployment.  It will cause ESPv2 to set scheme as "http" too even the backend connection is using TLS.  This will break the backend using IAP which requires scheme to be "https".

The changes: 
1)  remove the runtime flag, 
2) use SchemeHeaderOverride in the http_connection_manager to set scheme as "https" under following conditions:
* Deployed in server_less mode, 
* backend is grpc and uses TLS

### CODE REVIEW

File: docker/generic/start_proxy.py [hunk 1]
```diff
@@ -1356,6 +1356,7 @@ def gen_proxy_config(args):
     if args.on_serverless:
         proxy_conf.extend([
             "--compute_platform_override", SERVERLESS_PLATFORM])
+        proxy_conf.append("--server_less")
```
Comment (shuoyang2016) [inline docker/generic/start_proxy.py:1359]: Just use flag compute_platform_override?

File: src/go/configgenerator/listener_generator.go [hunk 2]
```diff
@@ -165,6 +166,33 @@ func MakeListener(serviceInfo *sc.ServiceInfo, filterGenerators []*filterconfig.
 	return listener, nil
 }
 
+// To fix b/221072669: a hack to work around b/221308324 where
+// Cloud Run always set :scheme header to http when using http2 protocol for grpc.
+// Override scheme header to https when following conditions meet:
+// * Deployed in serverless platform.
+// * Backend uses grpc
+// * All remote backends are using TLS
+func makeSchemeHeaderOverride(serviceInfo *sc.ServiceInfo) *corepb.SchemeHeaderTransformation {
```
Comment (shuoyang2016) [inline src/go/configgenerator/listener_generator.go:175]: Add a unit test?

Comment (qiwzhang) [reply src/go/configgenerator/listener_generator.go:175]: Done

File: docker/generic/start_proxy.py (hunk 1 above)
Comment (qiwzhang) [reply docker/generic/start_proxy.py]: Done

File: src/go/configgenerator/listener_generator.go [hunk 3]
```diff
@@ -165,6 +166,33 @@ func MakeListener(serviceInfo *sc.ServiceInfo, filterGenerators []*filterconfig.
...
+	if serviceInfo.Options.ComputePlatformOverride != util.ServerlessPlatform || !serviceInfo.GrpcSupportRequired {
+		return nil
+	}
+	allTLS := true
+	for _, v := range serviceInfo.RemoteBackendClusters {
+		if !v.UseTLS {
+			allTLS = false
```
Comment (shuoyang2016) [inline src/go/configgenerator/listener_generator.go:182]: nit: break

**Comment by qiwzhang** [comment]
PTAL

File: src/go/configgenerator/listener_generator.go [hunk 4]
```diff
@@ -165,6 +166,33 @@ func MakeListener(serviceInfo *sc.ServiceInfo, filterGenerators []*filterconfig.
...
+// Override scheme header to https when following conditions meet:
+// * Deployed in serverless platform.
+// * Backend uses grpc
+// * All remote backends are using TLS
+func makeSchemeHeaderOverride(serviceInfo *sc.ServiceInfo) *corepb.SchemeHeaderTransformation {
+	if serviceInfo.Options.ComputePlatformOverride != util.ServerlessPlatform || !serviceInfo.GrpcSupportRequired {
+		return nil
+	}
+	allTLS := true
+	for _, v := range serviceInfo.RemoteBackendClusters {
```
Comment (shuoyang2016) [inline src/go/configgenerator/listener_generator.go:180]: What will happen if some backends are TLS but some aren't?

Comment (qiwzhang) [reply src/go/configgenerator/listener_generator.go:180]: We could not use scheme_override.  For now, envoy scheme_override applies to all backends.  if some backend is not ssl,  the override will be wrong. 

I think this is ok.  All backends for serverless are using TLS.

Comment (shuoyang2016) [reply src/go/configgenerator/listener_generator.go:180]: > All backends for serverless are using TLS.

If this is the case, I think we should just return error here so that it simply fails at startup?

Comment (qiwzhang) [reply src/go/configgenerator/listener_generator.go:180]: That will be too brutal. In the case of not all backends use TLS,  my pr just don't do "override_scheme=https" ,  your proposal is, not to support it at all.  

I am afraid that there are some users using ESPv2 in such way.  We don't want to break them.  They may not need correct "scheme" header.   Only IAP backends require correct "scheme" headers.

**Review by TAOXUY — APPROVED** [review]

File: src/go/configgenerator/listener_generator.go (hunk 4 above)
Comment (shuoyang2016) [reply src/go/configgenerator/listener_generator.go:180]: I see. Where I come from is that it looks to me there are some combinations that we just cannot support. If there are some backends require https like IAP and some backends require http? Anyway we'll have to break one of them?

Comment (TAOXUY) [reply src/go/configgenerator/listener_generator.go:180]: Just discussed with Wayne about this. Since Envoy doesn't allow per backend scheme override, it would be a breaking change for certain use cases. Now, we only override scheme when makeSchemeHeaderOverride **all** backends are https which breaks the case Shuo mentioned where some backends are using IAP and some backends are HTTP.
. For fixing that, probably we can change to override scheme when **any** backend is https. Most Cloud Run backend should be https and the http backend should be fine with this wrong scheme header.

Comment (shuoyang2016) [reply src/go/configgenerator/listener_generator.go:180]: >Most Cloud Run backend should be https and the http backend should be fine with this wrong scheme header
Thanks. This sounds good.

**Review by shuoyang2016 — APPROVED** [review]

File: src/go/configgenerator/listener_generator.go (hunk 4 above)
Comment (qiwzhang) [reply src/go/configgenerator/listener_generator.go:180]: Code is updated to:  if ANY of backends is using SSL,  we will use scheme_override.  This is to supported multiple backends with mixed SSL and non-SSL.

**Review by nareddyt — APPROVED** [review]
I guess this is too much of an edge case to have an integration test for, right?

Can we somehow use HTTP echo server (with dynamic routing) to check it receives the correct `:scheme` header? 

https://github.com/GoogleCloudPlatform/esp-v2/blob/85785b79c3b719145e0b1779147ae4cca3c2aede/tests/endpoints/echo/server/app.go#L68

Maybe not worth it as long as Cloud Run e2e tests catch it
